### PR TITLE
fix: adjust cache_position handling with attention_mask

### PR DESCRIPTION
**Pull request by Krish0909**
What does this PR do?

This PR fixes an issue with incorrect masking behavior in the model.

The fix ensures that masked tokens are correctly handled during training and inference.

Fixes #40214

Motivation
Previously, the masking logic led to unintended behavior, which could cause incorrect attention patterns. This PR introduces the proper masking mechanism to align with expected model functionality.

Before submitting
 This PR fixes a bug (masking issue).
 I have read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#create-a-pull-request).

Who can review?
@ArthurZucker (text models)
@Rocketknight1 (trainer / general review)

### CODE REVIEW

File: src/transformers/masking_utils.py [hunk 1]
```diff
@@ -960,6 +960,14 @@ def create_chunked_causal_mask(
     )
     if early_exit:
         return attention_mask
+    
+    if attention_mask is not None and attention_mask.dim() == 2:
+        first_valid = (attention_mask > 0).int().argmax(dim=-1)  
+        max_shift = first_valid.min().item()
```
Comment (vasqu) [inline src/transformers/masking_utils.py:966]: Is this `.item()` call necessary? I'd rather avoid a cuda sync if possible

**Review by vasqu — COMMENTED** [review]
Had a feeling there was something as well. Also had some script to check left padding but never followed through :eyes: so good to see this! :D 

I'm not sure if we could move the logic to the mask creation instead, i.e. here https://github.com/huggingface/transformers/blob/54976714b2af440f7498bbba7a89f30980647c7b/src/transformers/masking_utils.py#L90-L99

Other than that, we would need a test to confirm these changes work as expected (looks like a small test in llama4).

cc @Cyrilvallez as you worked on the masks the most

**Comment by Cyrilvallez** [comment]
I don't understand this change, and how it would fix the issue mentioned. I believe this is wrong. Can you elaborate a bit more?

**Comment by Cyrilvallez** [comment]
I merged a fix for the linked issue here https://github.com/huggingface/transformers/pull/40324 👌
You can check it out and close this one! 🤗
